build-conventions/src/main/java/org/elasticsearch/gradle/internal/conventions/precommit/LicenseHeadersTask.java

```diff
@@ -0,0 +1,271 @@
+
+        // BSD 4-clause stuff (is disallowed below)
+        // we keep this here, in case someone adds BSD code for some reason, it should never be allowed.
+        matchers.add(subStringMatcher("BSD4 ", "Original BSD License (with advertising clause)", "All advertising materials"));
+        // Apache
+        matchers.add(subStringMatcher("AL   ", "Apache", "Licensed to Elasticsearch B.V. under one or more contributor"));
+        // Apache lz4-java
+        matchers.add(subStringMatcher("ALLZ4", "Apache LZ4-Java", "Copyright 2020 Adrien Grand and the lz4-java contributors"));
+        // Generated resources
+        matchers.add(subStringMatcher("GEN  ", "Generated", "ANTLR GENERATED CODE"));
+        // Vendored Code
+        matchers.add(subStringMatcher("VEN  ", "Vendored", "@notice"));
+
+        additionalLicenses.get().forEach(l ->
+            matchers.add(subStringMatcher(l.licenseFamilyCategory, l.licenseFamilyName, l.substringPattern))
+        );
+
+        reportConfiguration.setHeaderMatcher(new HeaderMatcherMultiplexer(matchers.toArray(IHeaderMatcher[]::new)));
+        reportConfiguration.setApprovedLicenseNames(approvedLicenses.stream().map(license -> {
+            SimpleLicenseFamily simpleLicenseFamily = new SimpleLicenseFamily();
+            simpleLicenseFamily.setFamilyName(license);
+            return simpleLicenseFamily;
+        }).toArray(SimpleLicenseFamily[]::new));
+
+        ClaimStatistic stats = generateReport(reportConfiguration, getReportFile());
+        boolean unknownLicenses = stats.getNumUnknown() > 0;
+        boolean unApprovedLicenses = stats.getNumUnApproved() > 0;
+        if (unknownLicenses || unApprovedLicenses) {
+            getLogger().error("The following files contain unapproved license headers:");
+            unapprovedFiles(getReportFile()).stream().forEachOrdered(unapprovedFile -> getLogger().error(unapprovedFile));
+            throw new GradleException("Check failed. License header problems were found. Full details: " + reportFile.getAbsolutePath());
+        }
+    }
+
+    private IHeaderMatcher subStringMatcher(String licenseFamilyCategory, String licenseFamilyName, String substringPattern) {
+        SubstringLicenseMatcher substringLicenseMatcher = new SubstringLicenseMatcher();
+        substringLicenseMatcher.setLicenseFamilyCategory(licenseFamilyCategory);
+        substringLicenseMatcher.setLicenseFamilyName(licenseFamilyName);
+        SubstringLicenseMatcher.Pattern pattern = new SubstringLicenseMatcher.Pattern();
+        pattern.setSubstring(substringPattern);
+        substringLicenseMatcher.addConfiguredPattern(pattern);
+        return substringLicenseMatcher;
+    }
+
+    private ClaimStatistic generateReport(ReportConfiguration config, File xmlReportFile) {
+        try {
+            Files.deleteIfExists(reportFile.toPath());
+            BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(xmlReportFile));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *bufferedWriter*. The resource is referenced at lines 210, 211. The resource is closed at line 229. There are other execution paths that do not contain closure statements, for example, when ClaimStatistic constructor (in method toXmlReportFile) throws an exception. Ensure *bufferedWriter* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).